File: src/main/java/dev/manideeplanka/bookerang/repositories/UserRepository.java
package dev.manideeplanka.bookerang.repositories;

import dev.manideeplanka.bookerang.common.DatabaseConfig;
import dev.manideeplanka.bookerang.models.User;
import org.jdbi.v3.core.Jdbi;

import java.util.Optional;

public class UserRepository {

    Jdbi jdbi;

    public UserRepository() {
        this.jdbi = DatabaseConfig.getJdbi();
    }


    public Optional<User> findByUsername(String username) throws Exception {
        return jdbi.withHandle(handle ->
                handle.createQuery("SELECT * FROM users WHERE username = :username").bind("username", username).mapToBean(User.class).findOne()
        );
    }


    public void signup(User user) throws Exception {
        jdbi.withHandle(handle -> {
                    handle.createUpdate("""
                                    INSERT INTO users (first_name, last_name, username, password, location)
                                    VALUES (
                                        :firstName,
                                        :lastName,
                                        :username,
                                        :password,
                                        ST_SetSRID(ST_MakePoint(:longitude, :latitude), 4326)::geography
                                    )
                                    """)
                            .bind("username", user.getUsername())
                            .bind("password", user.getPassword())
                            .bind("firstName", user.getFirstName())
                            .bind("lastName", user.getLastName())
                            .bind("latitude", user.getLatitude())
                            .bind("longitude", user.getLongitude())
                            .execute();
                    return null; //only returning this to avoid the warning in IDE, no practical use
                }
        );
    }

}

File: internal/repository/user_repository.go
package repository

import (
	"context"
	"database/sql"
	"fmt"

	"bookerang/internal/models"
)

type UserRepository struct {
	DB *sql.DB
}

func NewUserRepository(db *sql.DB) *UserRepository {
	return &UserRepository{DB: db}
}

func (r *UserRepository) FindByUsername(ctx context.Context, username string) (models.User, bool, error) {
	row := r.DB.QueryRowContext(ctx, `
		SELECT username, password, first_name, last_name
		FROM users
		WHERE username = $1
	`, username)

	var user models.User
	if err := row.Scan(&user.Username, &user.Password, &user.FirstName, &user.LastName); err != nil {
		if err == sql.ErrNoRows {
			return models.User{}, false, nil
		}
		return models.User{}, false, fmt.Errorf("find user by username: %w", err)
	}
	return user, true, nil
}

func (r *UserRepository) CreateUser(ctx context.Context, user models.User) error {
	_, err := r.DB.ExecContext(ctx, `
		INSERT INTO users (first_name, last_name, username, password, location)
		VALUES ($1, $2, $3, $4, ST_SetSRID(ST_MakePoint($5, $6), 4326)::geography)
	`, user.FirstName, user.LastName, user.Username, user.Password, user.Longitude, user.Latitude)
	if err != nil {
		return fmt.Errorf("create user: %w", err)
	}
	return nil
}
